_drafts/conviva-dag-process-time-prod-issue-draft.md

```diff
@@ -0,0 +1,160 @@
+
+Looking carefully at the flamegraph generated during the incident, you can see a very high load for call paths involving `AtomicUsize::fetch_sub` which was being called from creating and dropping a `ReadGuard` in [`flashmap`](https://github.com/Cassy343/flashmap), which we were using as a concurrent hash map. This concurrent hash map was being used as a type registry which was globally shared amongst all DAG's across our system.
+
+```rust
+use flashmap::{ReadHandle, WriteHandle};
+
+pub struct TypeRegistry<C: Clone, const N: usize = DEFAULT_N> {
+    writer: Mutex<WriteHandle<ShortTypeId, TypeMetadata<C, N>, BuildNoHashHasher<ShortTypeId>>>,
+    pub(crate) reader: ReadHandle<ShortTypeId, TypeMetadata<C, N>, BuildNoHashHasher<ShortTypeId>>
+}
+```
+
+In the context of this, the earlier graph about the context switches spiking during the incident makes some sense. The `ReadGuard` in the hot path of the flamegraph was responsible for handling reads from various threads and each thread would increment and decrement the counter.
+
+Now, one important thing about the hashmap in the type registry is that it is *almost* read-only. That is, it is initialized with some types on start-up and then only updated when a new type is seen but that rarely ever happened. However, on a critical path, it would keep checking to see if the type was already registered which is where the atomic increments and decrements were occurring.
+
+Now, the question was what to do about this. In the [flashmap documentation](https://crates.io/crates/flashmap), the performance comparison shows that in the read-heavy scenario [`dashmap`](https://github.com/xacrimon/dashmap) performed better in terms of both latency & throughput. Unfortunately, replacing `flashmap` with `dashmap` did nothing to fix the performance problems. In fact, the flamegraphs turned out to be worse in the same situation with `dashmap`.
+
+![](/assets/img/conviva/rtve/perf_incident_traffic_dashmapv2.svg)
+
+Finally, we implemented an [ArcSwap](https://github.com/vorner/arc-swap) based solution and the flamegraph for that was significantly better and the CPU load dropped to 40% in the perf environment.
+
+```rust
+use arc_swap::ArcSwap;
+
+pub struct TypeRegistry<C: Clone, const N: usize = DEFAULT_N> {
+    pub(crate) types:
+        ArcSwap<HashMap<ShortTypeId, TypeMetadata<C, N>, BuildNoHashHasher<ShortTypeId>>>,
+}
+```
+
+## Post Mortem
+
+So, ArcSwap finally fixed the problem but it is worthwhile to understand why exactly that worked where concurrent hash maps had failed. It's worth repeating here again that our type registry, which is where the hash map was being used, was an *almost* read-only scenario.
+
+First, let's look a little more closely at how concurrent hash maps typically operate (with a little hand-waving for simplicity's sake) - they use a *single, shared* counter which tracks how many readers are currently operating and allow a single writer to operate, not unlike a typical `RWLock`. 
```

Comment:
I don't think this is true.  Different maps have different designs, most of them don't just have one single atomic but try to spread out the atomics amongst different parts of the map to maximize concurrency.  I think flash map does that too.  The problem is we have very few entries so every if they are spread out, it still results in lots of contention.

Reply:
@velvia Yeah, dashmap does this too with shards and each shard is a `RwLock<HashMap>` but I didn't think it was worth going into those details, which is why I put in the note about hand waving. 

But, I've gone into those details in 7b65223885134b32a8e1ae4e71fb18525647faa5